src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+use crate::edn::{Edn, Error, List, Map, Set, Vector};
+
+pub(crate) fn tokenize(edn: &str) -> std::str::Chars {
+    edn.chars()
+}
+
+pub(crate) fn parse(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    Ok(match c {
+        Some('[') => read_vec(chars)?,
+        Some('(') => read_list(chars)?,
+        Some('@') => read_set(chars)?,
+        Some('{') => read_map(chars)?,
+        edn => parse_edn(edn, chars)?,
+    })
+}
+
+pub(crate) fn parse_edn(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    match c {
+        Some('\"') => Ok(read_str(chars)),
+        Some(':') => Ok(read_key(chars)),
+        Some(n) if n.is_numeric() => Ok(read_number(n, chars)?),
+        Some('-') => Ok(read_number('-', chars)?),
+        Some('\\') => Ok(read_char(chars)?),
+        Some(b) if b == 't' || b == 'f' || b == 'n' => Ok(read_bool_or_nil(b, chars)?),
```

Comment:
Can't this be `Some('t') | Some('f') | Some('n')`?

I think ifs like this on match clauses are probably slower than the regular's Rust pattern matching (idk the internals, I might be wrong, but I think it is worth the try).

Also I think it is cleaner.

If you're worried about using the `c` variable later, you can use the match binding (`@`) like this:

```rust
Some(b) @ Some('t') | Some('f') | Some('n') => Ok(read_bool_or_nil(b, chars)?)
```

Reply:
Reference: https://doc.rust-lang.org/stable/rust-by-example/flow_control/match/binding.html

Reply:
It will look like this: 
```b @ Some('t') | b @ Some('f') | b @ Some('n') => Ok(read_bool_or_nil(b.unwrap(), chars)?),```

Reply:
With binding:
`[14.173 us 14.470 us 14.756 us]`

With if:
`[13.232 us 13.264 us 13.303 us]`

Reply:
Ohh that's sad :slightly_frowning_face: